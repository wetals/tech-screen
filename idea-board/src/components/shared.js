export const MAX_COUNT = 140;
export const STORE_REF = "clearscore_ideas";
