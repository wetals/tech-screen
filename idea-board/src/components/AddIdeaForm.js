import React from "react";
import { MAX_COUNT } from "./shared";

const AddIdeaForm = props => {
  let remainingChar = MAX_COUNT - props.formData.description.length;

  return (
    <form
      className="idea"
      onSubmit={event => props.handleUserFormSubmit(event)}
    >
      <div className="field title">
        <input
          name="title"
          onChange={props.handleFormChange}
          type="text"
          autoFocus
          placeholder="Title"
          value={props.formData.title}
        />
      </div>

      <div className="field description">
        <textarea
          name="description"
          onChange={props.handleFormChange}
          placeholder="Description"
          value={props.formData.description}
        />
        {remainingChar < 20 && (
          <span className="counter">
            {`${remainingChar} characters remaining`}
          </span>
        )}
      </div>
      <div className="action-button">
        <button type="submit">Add</button>
      </div>
    </form>
  );
};

export default AddIdeaForm;
