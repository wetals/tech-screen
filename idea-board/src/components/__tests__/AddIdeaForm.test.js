import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { MAX_COUNT } from "../shared";

import AddIdeaForm from "../AddIdeaForm";

const testData = {
  formData: {
    title: "test",
    description:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum pretium venenatis luctus. Phasellus eu arcu euismod neque"
  },
  handleUserFormSubmit: jest.fn(),
  handleFormChange: jest.fn()
};

describe("AddIdeaForm component", () => {
  it("renders properly", () => {
    const wrapper = shallow(<AddIdeaForm {...testData} />);
    const element = wrapper.find("form");

    expect(element.length).toBe(1);
    expect(element.find("input").get(0).props.name).toBe("title");
    expect(element.find("textarea").get(0).props.name).toBe("description");
    expect(element.find("button").length).toBe(1);
  });

  it("submits the form properly", () => {
    const wrapper = shallow(<AddIdeaForm {...testData} />);
    expect(testData.handleFormChange).toHaveBeenCalledTimes(0);
    expect(testData.handleUserFormSubmit).toHaveBeenCalledTimes(0);

    wrapper.find("input").simulate("change");
    expect(testData.handleFormChange).toHaveBeenCalledTimes(1);

    wrapper.find("form").simulate("submit", testData.formData);
    expect(testData.handleUserFormSubmit).toHaveBeenCalledWith(
      testData.formData
    );
    expect(testData.handleUserFormSubmit).toHaveBeenCalledTimes(1);
  });

  it("displays character limits for textarea", () => {
    const wrapper = shallow(<AddIdeaForm {...testData} />);
    const count = testData.formData.description.length;
    const element = wrapper.find(".counter");
    expect(element.length).toBe(1);
    expect(element.text()).toBe(`${MAX_COUNT - count} characters remaining`);
  });

  it("renders snapshot properly", () => {
    const tree = renderer.create(<AddIdeaForm {...testData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
