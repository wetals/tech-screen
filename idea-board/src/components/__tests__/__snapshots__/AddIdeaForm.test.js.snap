// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddIdeaForm component renders snapshot properly 1`] = `
<form
  className="idea"
  onSubmit={[Function]}
>
  <div
    className="field title"
  >
    <input
      name="title"
      onChange={[Function]}
      placeholder="Title"
      type="text"
      value="test"
    />
  </div>
  <div
    className="field description"
  >
    <textarea
      name="description"
      onChange={[Function]}
      placeholder="Description"
      value="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum pretium venenatis luctus. Phasellus eu arcu euismod neque"
    />
    <span
      className="counter"
    >
      15 characters remaining
    </span>
  </div>
  <div
    className="action-button"
  >
    <button
      type="submit"
    >
      Add
    </button>
  </div>
</form>
`;
